# Homework and bake-off: Sentiment analysis

In [1]:
__author__ = "Christopher Potts"
__version__ = "CS224u, Stanford, Spring 2021"

## Contents

1. [Overview](#Overview)
1. [Methodological note](#Methodological-note)
1. [Set-up](#Set-up)
1. [Train set](#Train-set)
1. [Dev sets](#Dev-sets)
1. [A softmax baseline](#A-softmax-baseline)
1. [RNNClassifier wrapper](#RNNClassifier-wrapper)
1. [Error analysis](#Error-analysis)
1. [Homework questions](#Homework-questions)
  1. [Token-level differences [1 point]](#Token-level-differences-[1-point])
  1. [Training on some of the bakeoff data [1 point]](#Training-on-some-of-the-bakeoff-data-[1-point])
  1. [A more powerful vector-averaging baseline [2 points]](#A-more-powerful-vector-averaging-baseline-[2-points])
  1. [BERT encoding [2 points]](#BERT-encoding-[2-points])
  1. [Your original system [3 points]](#Your-original-system-[3-points])
1. [Bakeoff [1 point]](#Bakeoff-[1-point])
1. [Submission Instruction](#Submission-Instruction)

## Overview

This homework and associated bakeoff are devoted to supervised sentiment analysis using the ternary (positive/negative/neutral) version of the Stanford Sentiment Treebank (SST-3) as well as a new dev/test dataset drawn from restaurant reviews. Our goal in introducing the new dataset is to push you to create a system that performs well in both the movie and restaurant domains.

The homework questions ask you to implement some baseline system, and the bakeoff challenge is to define a system that does well at both the SST-3 test set and the new restaurant test set. Both are ternary tasks, and our central bakeoff score is the mean of the macro-FI scores for the two datasets. This assigns equal weight to all classes and datasets regardless of size.

The SST-3 test set will be used for the bakeoff evaluation. This dataset is already publicly distributed, so we are counting on people not to cheat by developing their models on the test set. You must do all your development without using the test set at all, and then evaluate exactly once on the test set and turn in the results, with no further system tuning or additional runs. __Much of the scientific integrity of our field depends on people adhering to this honor code__. 

One of our goals for this homework and bakeoff is to encourage you to engage in __the basic development cycle for supervised models__, in which you

1. Design a new system. We recommend starting with something simple.
1. Use `sst.experiment` to evaluate your system, using random train/test splits initially.
1. If you have time, compare your system with others using `sst.compare_models` or `utils.mcnemar`. (For discussion, see [this notebook section](sst_02_hand_built_features.ipynb#Statistical-comparison-of-classifier-models).)
1. Return to step 1, or stop the cycle and conduct a more rigorous evaluation with hyperparameter tuning and assessment on the `dev` set.

[Error analysis](#Error-analysis) is one of the most important methods for steadily improving a system, as it facilitates a kind of human-powered hill-climbing on your ultimate objective. Often, it takes a careful human analyst just a few examples to spot a major pattern that can lead to a beneficial change to the feature representations.

## Methodological note

You don't have to use the experimental framework defined below (based on `sst`). The only constraint we need to place on your system is that it must have a `predict_one` method that can map directly from an example text to a prediction, and it must be able to make predictions without having any information beyond the text. (For example, it can't depend on knowing which task the text comes from.) See [the bakeoff section below](#Bakeoff-[1-point]) for examples of functions that conform to this specification.

## Set-up

See [the first notebook in this unit](sst_01_overview.ipynb#Set-up) for set-up instructions.

In [2]:
from collections import Counter
import random
import numpy as np
import os
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction import DictVectorizer
import torch
import torch.nn as nn
from torch_rnn_classifier import TorchRNNClassifier
from torch_tree_nn import TorchTreeNN
import sst
import sst_mod
from sklearn.metrics import classification_report
from iit import get_equality_dataset, get_IIT_equality_dataset, get_IIT_equality_dataset_both
from torch_deep_neural_classifier_iit import TorchDeepNeuralClassifierIIT, TorchDeepNeuralClassifier
from torch_rnn_classifier import TorchRNNClassifier
import utils

In [3]:
SST_HOME = os.path.join('data', 'sentiment')

## Train set

Our primary train set is the SST-3 train set:

In [4]:
sst_train = sst.train_reader(SST_HOME)

In [5]:
sst_train.shape[0]

8544

This is the train set we will use for all the regular homework questions. You are welcome to bring in new datasets for your original system. You are also free to add `include_subtrees=True`. This is very likely to lead to better systems, but it substantially increases the overall size of the dataset (from 8,544 examples to 159,274), which will in turn substantially increase the time it takes to run experiments.

See [this notebook](sst_01_overview.ipynb) for additional details of this dataset.

## Dev sets

We have two development set. SST3-dev consists of sentences from movie reviews, just like SST-3 train:

In [6]:
sst_dev = sst.dev_reader(SST_HOME)

Our new bakeoff dev set consists of sentences from restaurant reviews:

In [7]:
bakeoff_dev = sst.bakeoff_dev_reader(SST_HOME)

In [8]:
bakeoff_dev.sample(3, random_state=1).to_dict(orient='records')

[{'example_id': 57,
  'sentence': 'I would recommend that you make reservations in advance.',
  'label': 'neutral',
  'is_subtree': 0},
 {'example_id': 590,
  'sentence': 'We were welcomed warmly.',
  'label': 'positive',
  'is_subtree': 0},
 {'example_id': 1968,
  'sentence': 'We have been to Oceanaire twice in the last 6 weeks.',
  'label': 'neutral',
  'is_subtree': 0}]

Here is the label distribution:

In [9]:
bakeoff_dev.label.value_counts()

neutral     1019
positive     777
negative     565
Name: label, dtype: int64

The label distribution for the corresponding test set is similar to this.

## A softmax baseline

This example is here mainly as a reminder of how to use our experimental framework with linear models:

In [10]:
def unigrams_phi(text):
    return Counter(text.split())

Thin wrapper around `LogisticRegression` for the sake of `sst.experiment`:

In [11]:
def fit_softmax_classifier(X, y):
    mod = LogisticRegression(
        fit_intercept=True,
        solver='liblinear',
        multi_class='ovr')
    mod.fit(X, y)
    return mod

The experimental run with some notes:

In [12]:
softmax_experiment = sst.experiment(
    sst.train_reader(SST_HOME),   # Train on any data you like except SST-3 test!
    unigrams_phi,                 # Free to write your own!
    fit_softmax_classifier,       # Free to write your own!
    assess_dataframes=[sst_dev, bakeoff_dev]) # Free to change this during development!

Assessment dataset 1
              precision    recall  f1-score   support

    negative      0.628     0.689     0.657       428
     neutral      0.343     0.153     0.211       229
    positive      0.629     0.750     0.684       444

    accuracy                          0.602      1101
   macro avg      0.533     0.531     0.518      1101
weighted avg      0.569     0.602     0.575      1101

Assessment dataset 2
              precision    recall  f1-score   support

    negative      0.272     0.692     0.391       565
     neutral      0.429     0.113     0.179      1019
    positive      0.409     0.346     0.375       777

    accuracy                          0.328      2361
   macro avg      0.370     0.384     0.315      2361
weighted avg      0.385     0.328     0.294      2361

Mean of macro-F1 scores: 0.416


In [13]:
def one_hot(label):
    sents = ['positive', 'neutral', 'negative']
    return np.eye(len(sents))[sents.index(label)]

def build_dataset_subtrees(dataframes, phi, vectorizer=None, vectorize=True):
    if isinstance(dataframes, (list, tuple)):
        df = pd.concat(dataframes)
    else:
        df = dataframes

    raw_examples = list(df.sentence.values)

    # feat_dicts = list(df.left_label.apply(phi).values)
    left_labels = df.left_label.values
    right_labels = df.right_label.values

    feat_dicts = [np.concatenate((one_hot(left_labels[i]), one_hot(right_labels[i]))) for i in range(len(left_labels))]

    if 'sentence_label' in df.columns:
        labels = list(df.sentence_label.values)
    else:
        labels = None

    feat_matrix = None
    if vectorize:
        # In training, we want a new vectorizer:
        if vectorizer is None:
            vectorizer = DictVectorizer(sparse=False)
            feat_matrix = vectorizer.fit_transform(feat_dicts)
        # In assessment, we featurize using the existing vectorizer:
        else:
            feat_matrix = vectorizer.transform(feat_dicts)
    else:
        feat_matrix = feat_dicts

    return {'X': feat_matrix,
            'y': labels,
            'vectorizer': vectorizer,
            'raw_examples': raw_examples}

In [14]:
sentiment_iit_train_df = pd.read_csv('sst_tree_train.csv')
sentiment_iit_dev_df = pd.read_csv('sst_tree_dev.csv')

softmax_tree_experiment = sst_mod.experiment(
    sentiment_iit_train_df,
    unigrams_phi,
    fit_softmax_classifier,
    assess_dataframes=[sentiment_iit_dev_df],
    vectorize=False,
    build_dataset_fn=build_dataset_subtrees)

              precision    recall  f1-score   support

    negative      1.000     1.000     1.000       428
     neutral      1.000     1.000     1.000       229
    positive      1.000     1.000     1.000       444

    accuracy                          1.000      1101
   macro avg      1.000     1.000     1.000      1101
weighted avg      1.000     1.000     1.000      1101



In [15]:
class IIT_SentimentAnalysisDataset:
    """
    Create data as follows:

    (base sentence, source sentence, label of base, label of base intervened by source, location of intervention)

    where location of intervention is either LEFT_SUBTREE or RIGHT_SUBTREE, and
    label is either POSITIVE, NEUTRAL, or NEGATIVE.
    """

    LEFT_SUBTREE = 0
    RIGHT_SUBTREE = 1
    POSITIVE = 0
    NEUTRAL = 1
    NEGATIVE = 2
    LABELS = ['positive', 'neutral', 'negative']

    def one_hot(self, label):
        return np.eye(len(self.LABELS))[self.LABELS.index(label)]

    def __init__(self, subtree_df, root_model, variable) -> None:
        """
        subtree df has rows that give values for:
        (sentence, label of sentence, left tree, label of left tree, right tree, label of right tree)

        root_model is a function mapping from (left label, right label) --> full sentence label. 
        Ideally this is a simple model (e.g. LogisticRegression)

        variable is either LEFT_SUBTREE or RIGHT_SUBTREE
        """
        self.subtree_df = subtree_df
        self.root_model = root_model
        self.variable = variable

    def create_take_two(self):
        def intervene(base, source, location):
            interventions_indices = ['left_label', 'right_label']
            intervention_input = [0, 0] # set up two slots for values
            intervention_input[location] = self.one_hot(source[interventions_indices[location]])
            intervention_input[1 - location] = self.one_hot(base[interventions_indices[1 - location]])
            return np.concatenate(intervention_input)
        
        n = len(self.subtree_df)

        # data = [(self.subtree_df.iloc[b].sentence, 
        #          self.subtree_df.iloc[s].sentence, 
        #          self.subtree_df.iloc[b].sentence_label, 
        #          self.root_model.predict(intervene(self.subtree_df.iloc[b], self.subtree_df.iloc[s], self.variable))[0],
        #          self.variable) for b in range(n) for s in range(n)]
        base = [self.subtree_df.iloc[b].sentence for b in range(n) for _ in range(n)]
        source = [self.subtree_df.iloc[s].sentence for _ in range(n) for s in range(n)]
        y = [self.subtree_df.iloc[b].sentence_label for b in range(n) for _ in range(n)]
        IIT_y = self.root_model.predict([intervene(self.subtree_df.iloc[b], self.subtree_df.iloc[s], self.variable) for b in range(n) for s in range(n)])
        interventions = [self.variable] * (n * n)
        self.counts = Counter(zip(y, [self.subtree_df.iloc[s].sentence_label for _ in range(n) for s in range(n)]))
        print(self.counts)
        self.data = list(zip(base, source, y, IIT_y, interventions))
        # should I shuffle my data?
        random.shuffle(self.data)
        # why are we making a copy?
        data = self.data.copy()
        base, source, y, IIT_y, interventions = zip(*data)
        self.base = np.array(base)
        self.source = np.array(source)
        self.y = np.array(y)
        self.IIT_y = np.array(IIT_y)
        self.interventions = np.array(interventions)
        self.sources = [source] # feels like the same thing 
        return self.base, self.sources, self.y, self.IIT_y, self.interventions

    def create(self):
        self.counts = {}
        self.data = []
        for b in range(len(self.subtree_df)):
            for s in range(len(self.subtree_df)):
                base = self.subtree_df.iloc[b]
                source = self.subtree_df.iloc[s]
                base_label = base['sentence_label']
                # extract subtree labels for root model evaluation
                base_input = [base['left_label'], base['right_label']]
                source_input = [source['left_label'], source['right_label']]
                # intervene on base input with the source input at the intervention location
                base_input[self.variable] = source_input[self.variable]
                # run model to compute root value from leaves
                model_input = [np.concatenate([self.one_hot(label) for label in base_input])]
                iit_label = self.root_model.predict(model_input)[0]
                # BE SURE TO COUNT +, -, and neutrals!!!
                self.data.append((base.sentence, source.sentence, 
                                 base_label, iit_label, self.variable))
                self.counts[(base_label, iit_label)] = self.counts.get((base_label, iit_label), 0) + 1
        # should I shuffle my data?
        random.shuffle(self.data)
        # why are we making a copy?
        data = self.data.copy()
        base, source, y, IIT_y, interventions = zip(*data)
        self.base = np.array(base)
        self.source = np.array(source)
        self.y = np.array(y)
        self.IIT_y = np.array(IIT_y)
        self.interventions = np.array(interventions)
        self.sources = [source] # feels like the same thing 
        print(self.counts)
        return self.base, self.sources, self.y, self.IIT_y, self.interventions

LABELS = ['positive', 'neutral', 'negative']

def get_IIT_sentiment_dataset(subtree_df, root_model, variable, phi, vectorizer=None, vectorize=True):
    """
    variable is 0 or 1, corresponding to LEFT_SUBTREE or RIGHT_SUBTREE

    phi is function to extract features from sentence
    """
    dataset = IIT_SentimentAnalysisDataset(subtree_df, root_model, variable)
    X_base, X_sources, y_base, y_IIT, interventions = dataset.create()
    # use phi to extract features from sentences in base and source inputs
    X = [phi(x) for x in X_base] + [phi(x) for source in X_sources for x in source]
    if vectorize:
        if vectorizer is None:
            vectorizer = DictVectorizer(sparse=False)
            X = vectorizer.fit_transform(X)
        else:
            X = vectorizer.transform(X)
    # reshape X to list of length len(X_base)
    X = [X[i:i + len(X_base)] for i in range(0, len(X), len(X_base))]
    # first list is the base, the rest are the sources
    # should I convert these to tensors? seems that build_dataset takes care of this anyway?
    X_base = torch.tensor(X[0]) if vectorize else X[0]
    X_sources = [torch.tensor(source) if vectorize else source for source in X[1:]]
    y_base = torch.tensor([LABELS.index(label) for label in y_base])
    y_IIT = torch.tensor([LABELS.index(label) for label in y_IIT])
    interventions = torch.tensor(interventions)
    return X_base, X_sources, y_base, y_IIT, interventions, vectorizer

def get_IIT_sentiment_devset(dev_df, variable, phi, vectorizer, vectorize=True):
    LABELS = ['positive', 'neutral', 'negative']
    sentences = [phi(x) for x in dev_df['sentence'].values]
    if vectorize:
        sentences = torch.tensor(vectorizer.transform(sentences))
    labels = torch.tensor([LABELS.index(x) for x in dev_df['label'].values])
    # can I make my sources and IIT labels empty??
    # what if I don't have a way to measure IIT labels for the test set?
    sources = [sentences]
    y_IIT = labels
    interventions = torch.tensor([variable] * len(dev_df))
    return sentences, sources, labels, y_IIT, interventions

In [16]:
LEFT = 0
RIGHT = 1
softmax_root_model = softmax_tree_experiment['model']

data_size = 10
X_base, X_sources, y_base, y_IIT, interventions, vectorizer = get_IIT_sentiment_dataset(sentiment_iit_dev_df.sample(data_size), softmax_root_model, LEFT, unigrams_phi)

split = (data_size * data_size) // 5

X_base_train = X_base[split:]
X_sources_train = [source[split:] for source in X_sources]
y_base_train = y_base[split:]
y_IIT_train = y_IIT[split:]
interventions_train = interventions[split:]


print(X_base_train.shape)

{('negative', 'negative'): 56, ('negative', 'neutral'): 3, ('negative', 'positive'): 1, ('neutral', 'negative'): 18, ('neutral', 'neutral'): 9, ('neutral', 'positive'): 3, ('positive', 'negative'): 6, ('positive', 'neutral'): 3, ('positive', 'positive'): 1}
torch.Size([80, 135])


In [17]:
embedding_dim = X_base.shape[1]
V1 = 0
V2 = 1
both = 2
# similar to our alignment in the IIT accuracy section?
# aligning V1 to left side of layer 1, and V2 to the right side
# we are defining both as a list with two values -- why not encode it as a single range from 0  to dim * 2?
id_to_coords = {V1:{1: [{"layer":1, "start":0, "end":embedding_dim}]}, \
    V2: {1: [{"layer":1, "start":embedding_dim, "end":embedding_dim*2}]}, \
    both: {1: [{"layer":1, "start":0, "end":embedding_dim},{"layer":1, "start":embedding_dim, "end":embedding_dim*2}]}}

# gives back an IIT dataset based off of the Premack dataset, coming up with 
# all possible permutations of same/different shape pairs and same/different base-source pairs?
# X_base_train, X_sources_train, y_base_train, y_IIT_train, interventions = get_IIT_equality_dataset("V1", embedding_dim ,data_size)

# this is a different model from the one we defined in the previous cell, but with a similar idea?
model = TorchDeepNeuralClassifierIIT(hidden_dim=embedding_dim*4, hidden_activation=torch.nn.ReLU(), num_layers=3, id_to_coords=id_to_coords)
# model.fit() function internally calls on model.create_dataset(), which creates dataset in a way that pairs off
# source and base inputs?
_ = model.fit(X_base_train, X_sources_train, y_base_train, y_IIT_train, interventions_train)

# this is a runtime error I've also encountered in antra (with no change to the original code)
# could this be due to mismatching pytorch versions??

Stopping after epoch 54. Training loss did not improve more than tol=1e-05. Final error is 4.617796093953075e-06.

In [18]:
X_base_test = X_base[:split]
X_sources_test = [source[:split] for source in X_sources]
y_base_test = y_base[:split]
y_IIT_test = y_IIT[:split]
interventions_test = interventions[:split]

IIT_preds, base_preds = model.model(model.prep_input(X_base_test, X_sources_test, interventions_test))
IIT_preds = np.array(IIT_preds.argmax(axis=1).cpu())
base_preds = np.array(base_preds.argmax(axis=1).cpu())
print(classification_report(y_base_test, base_preds))
print(classification_report(y_IIT_test, IIT_preds))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         3
           1       1.00      1.00      1.00         7
           2       1.00      1.00      1.00        10

    accuracy                           1.00        20
   macro avg       1.00      1.00      1.00        20
weighted avg       1.00      1.00      1.00        20

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         2
           1       1.00      1.00      1.00         4
           2       1.00      1.00      1.00        14

    accuracy                           1.00        20
   macro avg       1.00      1.00      1.00        20
weighted avg       1.00      1.00      1.00        20



In [19]:
X_base_dev, X_sources_dev, y_base_dev, y_IIT_dev, interventions_dev = get_IIT_sentiment_devset(bakeoff_dev, LEFT, unigrams_phi, vectorizer)

IIT_preds, base_preds = model.model(model.prep_input(X_base_dev, X_sources_dev, interventions_dev))
base_preds = np.array(base_preds.argmax(axis=1).cpu())
print(classification_report(y_base_dev, base_preds))

              precision    recall  f1-score   support

           0       0.25      0.05      0.08       777
           1       0.00      0.00      0.00      1019
           2       0.24      0.94      0.38       565

    accuracy                           0.24      2361
   macro avg       0.16      0.33      0.15      2361
weighted avg       0.14      0.24      0.12      2361



In [26]:
from torch_bert_classifier import HfBertClassifierModel

test_bert = HfBertClassifierModel(3)

# test_data = HfBertClassifier().build_dataset(["I hated everything about this"])
# X = torch.cat(test_data.tensors, dim=1)

c = list(test_bert.layers)
print(len(c))
c

5


[BertEmbeddings(
   (word_embeddings): Embedding(28996, 768, padding_idx=0)
   (position_embeddings): Embedding(512, 768)
   (token_type_embeddings): Embedding(2, 768)
   (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
   (dropout): Dropout(p=0.1, inplace=False)
 ),
 BertEncoder(
   (layer): ModuleList(
     (0): BertLayer(
       (attention): BertAttention(
         (self): BertSelfAttention(
           (query): Linear(in_features=768, out_features=768, bias=True)
           (key): Linear(in_features=768, out_features=768, bias=True)
           (value): Linear(in_features=768, out_features=768, bias=True)
           (dropout): Dropout(p=0.1, inplace=False)
         )
         (output): BertSelfOutput(
           (dense): Linear(in_features=768, out_features=768, bias=True)
           (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
           (dropout): Dropout(p=0.1, inplace=False)
         )
       )
       (intermediate): BertIntermediate(
     

In [25]:
# test_model = TorchBertClassifierIIT(weights_name='bert-base-cased')
# test_model.n_classes_ = 3
# bert_iit = test_model.build_graph()
LEFT = 0
RIGHT = 1
BOTH = 2
dim = 768
half_dim = dim // 2

id_to_coords = {LEFT:{3: [{"layer":0, "start":0, "end":half_dim}]}, \
    RIGHT: {3: [{"layer":0, "start":half_dim, "end":dim}]}, \
    BOTH: {3: [{"layer":0, "start":0, "end":half_dim},{"layer":0, "start":half_dim, "end":dim}]}}
test_model = TorchBertClassifierIIT(id_to_coords)
base = ['I hate you!', 'I love how awesome you are']
sources = [base]
coord_ids = [0, 0]
y = [0, 1]
y_IIT = [0, 1]

_ = test_model.fit(base, sources, y, y_IIT, coord_ids)

# for k in range(len(model.layers)):
#     # build a get request for which layer and which values we want to see
#     get_coord = {"layer":k, "start":0, "end":embedding_dim*4}
#     print(f"\nLayer {k}:", model.layers[k])
#     # prints actual hidden values (activations) of the neural network layer
#     print("\nNeural Activations:", model.retrieve_activations(torch.tensor([]), get_coord, None))

Finished epoch 9 of 1000; error is 5.2372517585754395

KeyboardInterrupt: 

In [1]:
base = ['love love love love']
sources = [base]
coord_ids = [0]

LABELS = ['positive', 'negative']
_, y_ = test_model.model(test_model.prep_input(base, sources, coord_ids))
y_ = np.array(y_.argmax(axis=1).cpu())
[LABELS[y] for y in y_]

NameError: name 'test_model' is not defined

`softmax_experiment` contains a lot of information that you can use for error analysis; see [this section below](#Error-analysis) for starter code.

## RNNClassifier wrapper

This section illustrates how to use `sst.experiment` with `TorchRNNClassifier`.

To featurize examples for an RNN, we can just get the words in order, letting the model take care of mapping them into an embedding space.

In [ ]:
def rnn_phi(text):
    return text.split()

The model wrapper gets the vocabulary using `sst.get_vocab`. If you want to use pretrained word representations in here, then you can have `fit_rnn_classifier` build that space too; see [this notebook section for details](sst_03_neural_networks.ipynb#Pretrained-embeddings). See also [torch_model_base.py](torch_model_base.py) for details on the many optimization parameters that `TorchRNNClassifier` accepts.

In [ ]:
def fit_rnn_classifier(X, y):
    sst_glove_vocab = utils.get_vocab(X, mincount=2)
    mod = TorchRNNClassifier(
        sst_glove_vocab,
        early_stopping=True)
    mod.fit(X, y)
    return mod

In [ ]:
rnn_experiment = sst.experiment(
    sst.train_reader(SST_HOME),
    rnn_phi,
    fit_rnn_classifier,
    vectorize=False,  # For deep learning, use `vectorize=False`.
    assess_dataframes=[sst_dev, bakeoff_dev])

## Error analysis

This section begins to build an error-analysis framework using the dicts returned by `sst.experiment`. These have the following structure:

```
'model': trained model
'phi': the feature function used
'train_dataset':
   'X': feature matrix
   'y': list of labels
   'vectorizer': DictVectorizer,
   'raw_examples': list of raw inputs, before featurizing   
'assess_datasets': list of datasets, each with the same structure as the value of 'train_dataset'
'predictions': list of lists of predictions on the assessment datasets
'metric': `score_func.__name__`, where `score_func` is an `sst.experiment` argument
'score': the `score_func` score on the each of the assessment dataasets
```
The following function just finds mistakes, and returns a `pd.DataFrame` for easy subsequent processing:

In [ ]:
def find_errors(experiment):
    """Find mistaken predictions.

    Parameters
    ----------
    experiment : dict
        As returned by `sst.experiment`.

    Returns
    -------
    pd.DataFrame

    """
    dfs = []
    for i, dataset in enumerate(experiment['assess_datasets']):
        df = pd.DataFrame({
            'raw_examples': dataset['raw_examples'],
            'predicted': experiment['predictions'][i],
            'gold': dataset['y']})
        df['correct'] = df['predicted'] == df['gold']
        df['dataset'] = i
        dfs.append(df)
    return pd.concat(dfs)

In [ ]:
softmax_analysis = find_errors(softmax_experiment)

In [ ]:
rnn_analysis = find_errors(rnn_experiment)

Here we merge the sotmax and RNN experiments into a single DataFrame:

In [ ]:
analysis = softmax_analysis.merge(
    rnn_analysis, left_on='raw_examples', right_on='raw_examples')

analysis = analysis.drop('gold_y', axis=1).rename(columns={'gold_x': 'gold'})

The following code collects a specific subset of examples; small modifications to its structure will give you different interesting subsets:

In [ ]:
# Examples where the softmax model is correct, the RNN is not,
# and the gold label is 'positive'

error_group = analysis[
    (analysis['predicted_x'] == analysis['gold'])
    &
    (analysis['predicted_y'] != analysis['gold'])
    &
    (analysis['gold'] == 'positive')
]

In [ ]:
error_group.shape[0]

In [ ]:
for ex in error_group['raw_examples'].sample(5, random_state=1):
    print("="*70)
    print(ex)

## Homework questions

Please embed your homework responses in this notebook, and do not delete any cells from the notebook. (You are free to add as many cells as you like as part of your responses.)

### Token-level differences [1 point]

We can begin to get a sense for how our two dev sets differ by considering the most frequent tokens from each. This question asks you to begin such analysis.

Your task: write a function `get_token_counts` that, given a `pd.DataFrame` in the format of our datasets, tokenizes the example sentences based on whitespace and creates a count distribution over all of the tokens. The function should return a `pd.Series` sorted by frequency; if you create a count dictionary `d`, then `pd.Series(d).sort_values(ascending=False)` will give you what you need.

In [ ]:
def get_token_counts(df):
    pass
    ##### YOUR CODE HERE



In [ ]:
def test_get_token_counts(func):
    df = pd.DataFrame([
        {'sentence': 'a a b'},
        {'sentence': 'a b a'},
        {'sentence': 'a a a b.'}])
    result = func(df)
    for token, expected in (('a', 7), ('b', 2), ('b.', 1)):
        actual = result.loc[token]
        assert actual == expected, \
            "For token {}, expected {}; got {}".format(
            token, expected, actual)

In [ ]:
if 'IS_GRADESCOPE_ENV' not in os.environ:
    test_get_token_counts(get_token_counts)

As you develop your original system, you might review these results. The two dev sets have different vocabularies and different low-level encoding details that are sure to impact model performance, especially when one considers that the train set is like `sst_dev` in all these respects. For additional discussion, see [this notebook section](sst_01_overview.ipynb#Tokenization).

### Training on some of the bakeoff data [1 point]

We have so far presented the bakeoff dev set as purely for evaluation. Since the train set consists entirely of SST-3 data, this makes the bakeoff split especially challenging. We might be able to reduce the challenging by adding some of the bakeoff dev set to the train set, keeping some of it for evaluation. The current question asks to begin explore the effects of such training.

Your task: write a function `run_mixed_training_experiment`. The function should:

1. Take as inputs (a) a model training wrapper like `fit_softmax_classifier` and (b) an integer `bakeoff_train_size` specifying the number of examples from `bakeoff_dev` that should be included in the train set.
1. Split `bakeoff_dev` so that the first `bakeoff_train_size` examples are in the train set and the rest are used for evaluation.
1. Use `sst.experiment` with the user-supplied model training wrapper, `unigram_phi` as defined above, and a train set that consists of SST-3 train and the train portion of `bakeoff_dev` as defined in step 2. The value of `assess_dataframes` should be a list consisting of the SST-3 dev set and the evaluation portion of `bakeoff_dev` as defined in step 2.
1. Return the return value of `sst.experiment`.

The function `test_run_mixed_training_experiment` will help you iterate to the required design.

In [ ]:
def run_mixed_training_experiment(wrapper_func, bakeoff_train_size):
    pass
    ##### YOUR CODE HERE



In [ ]:
def test_run_mixed_training_experiment(func):
    bakeoff_train_size = 1000
    experiment = func(fit_softmax_classifier, bakeoff_train_size)

    assess_size = len(experiment['assess_datasets'])
    assert len(experiment['assess_datasets']) == 2, \
        ("The evaluation should be done on two datasets: "
         "SST3 and part of the bakeoff dev set. "
         "You have {} datasets.".format(assess_size))

    bakeoff_test_size = bakeoff_dev.shape[0] - bakeoff_train_size
    expected_eval_examples = bakeoff_test_size + sst_dev.shape[0]
    eval_examples = sum(len(d['raw_examples']) for d in experiment['assess_datasets'])
    assert expected_eval_examples == eval_examples, \
        "Expected {} evaluation examples; got {}".format(
        expected_eval_examples, eval_examples)

In [ ]:
if 'IS_GRADESCOPE_ENV' not in os.environ:
    test_run_mixed_training_experiment(run_mixed_training_experiment)

### A more powerful vector-averaging baseline [2 points]

In [Distributed representations as features](sst_03_neural_networks.ipynb#Distributed-representations-as-features), we looked at a baseline for the ternary SST-3 problem in which each example is modeled as the mean of its GloVe representations. A `LogisticRegression` model was used for prediction. A neural network might do better with these representations, since there might be complex relationships between the input feature dimensions that a linear classifier can't learn. To address this question, we want to get set up to run the experiment with a shallow neural classifier. 

Your task: write and submit a model wrapper function around `TorchShallowNeuralClassifier`. This function should implement hyperparameter search according to this specification:

* Set `early_stopping=True` for all experiments.
* Using 3-fold cross-validation, exhaustively explore this set of hyperparameter combinations:
  * The hidden dimensionality at 50, 100, and 200.
  * The hidden activation function as `nn.Tanh()` and `nn.ReLU()`.
* For all other parameters to `TorchShallowNeuralClassifier`, use the defaults.

See [this notebook section](sst_02_hand_built_features.ipynb#Hyperparameter-search) for examples. You are not required to run a full evaluation with this function using `sst.experiment`, but we assume you will want to.

We're not evaluating the quality of your model. (We've specified the protocols completely, but there will still be variation in the results.) However, the primary goal of this question is to get you thinking more about this strong baseline feature representation scheme for SST-3, so we're sort of hoping you feel compelled to try out variations on your own.

In [ ]:
from torch_shallow_neural_classifier import TorchShallowNeuralClassifier

def fit_shallow_neural_classifier_with_hyperparameter_search(X, y):
    pass
    ##### YOUR CODE HERE


### BERT encoding [2 points]

We might hypothesize that encoding our examples with BERT will yield improvements over the GloVe averaging method explored in the previous question, since BERT implements a much more complex and data-driven function for this kind of combination. This question asks you to begin exploring this general hypothesis.

Your task: write a function `hf_cls_phi` that uses Hugging Face functionality to encode individual examples with BERT and returns the final output representation above the [CLS] token.

You are not required to evaluate this feature function, but it is easy to do so with `sst.experiment` and `vectorize=False` (since your feature function directly encodes every example as a vector). Your code should also be a natural basis for even more powerful approaches – for example, it might be even better to pool all the output states rather than using just the first output state. Another option is [fine-tuning](finetuning.ipynb).

In [ ]:
from transformers import BertModel, BertTokenizer
import vsm

# Instantiate a Bert model and tokenizer based on `bert_weights_name`:
bert_weights_name = 'bert-base-uncased'
##### YOUR CODE HERE


def hf_cls_phi(text):
    # Get the ids. `vsm.hf_encode` will help; be sure to
    # set `add_special_tokens=True`.
    ##### YOUR CODE HERE


    # Get the BERT representations. `vsm.hf_represent` will help:
    ##### YOUR CODE HERE


    # Index into `reps` to get the representation above [CLS].
    # The shape of `reps` should be (1, n, 768), where n is the
    # number of tokens. You need the 0th element of the 2nd dim:
    ##### YOUR CODE HERE


    # These conversions should ensure that you can work with the
    # representations flexibly. Feel free to change the variable
    # name:
    return cls_rep.cpu().numpy()

In [ ]:
def test_hf_cls_phi(func):
    rep = func("Just testing!")

    expected_shape = (768,)
    result_shape = rep.shape
    assert rep.shape == (768,), \
        "Expected shape {}; got {}".format(
        expected_shape, result_shape)

    # String conversion to avoid precision errors:
    expected_first_val = str(0.1709)
    result_first_val = "{0:.04f}".format(rep[0])

    assert expected_first_val == result_first_val, \
        ("Unexpected representation values. Expected the "
        "first value to be {}; got {}".format(
            expected_first_val, result_first_val))

In [ ]:
if 'IS_GRADESCOPE_ENV' not in os.environ:
    test_hf_cls_phi(hf_cls_phi)

Note: encoding all of SST-3 train (no subtrees) takes about 11 minutes on my 2015 iMac, CPU only (32GB).

### Your original system [3 points]

Your task is to develop an original model for the SST-3 problem and our new bakeoff dataset. There are many options. If you spend more than a few hours on this homework problem, you should consider letting it grow into your final project! Here are some relatively manageable ideas that you might try:

1. We didn't systematically evaluate the `bidirectional` option to the `TorchRNNClassifier`. Similarly, that model could be tweaked to allow multiple LSTM layers (at present there is only one), and you could try adding layers to the classifier portion of the model as well.

1. We've already glimpsed the power of rich initial word representations, and later in the course we'll see that smart initialization usually leads to a performance gain in NLP, so you could perhaps achieve a winning entry with a simple model that starts in a great place.

1. Our [practical introduction to contextual word representations](finetuning.ipynb) covers pretrained representations and interfaces that are likely to boost the performance of any system.

We want to emphasize that this needs to be an __original__ system. It doesn't suffice to download code from the Web, retrain, and submit. You can build on others' code, but you have to do something new and meaningful with it. See the course website for additional guidance on how original systems will be evaluated.

In the cell below, please provide a brief technical description of your original system, so that the teaching team can gain an understanding of what it does. This will help us to understand your code and analyze all the submissions to identify patterns and strategies.  We also ask that you report the best score your system got during development (your best average of macro-F1 scores), just to help us understand how systems performed overall.

<font color='red'>Please review the descriptions in the following comment and follow the instructions.</font>

In [ ]:
# PLEASE MAKE SURE TO INCLUDE THE FOLLOWING BETWEEN THE START AND STOP COMMENTS:
#   1) Textual description of your system.
#   2) The code for your original system.
#   3) The score achieved by your system in place of MY_NUMBER.
#        With no other changes to that line.
#        You should report your score as a decimal value <=1.0
# PLEASE MAKE SURE NOT TO DELETE OR EDIT THE START AND STOP COMMENTS

# NOTE: MODULES, CODE AND DATASETS REQUIRED FOR YOUR ORIGINAL SYSTEM
# SHOULD BE ADDED BELOW THE 'IS_GRADESCOPE_ENV' CHECK CONDITION. DOING
# SO ABOVE THE CHECK MAY CAUSE THE AUTOGRADER TO FAIL.

# START COMMENT: Enter your system description in this cell.
# My peak score was: MY_NUMBER
if 'IS_GRADESCOPE_ENV' not in os.environ:
    pass

# STOP COMMENT: Please do not remove this comment.

## Bakeoff [1 point]

As we said above, the bakeoff evaluation data is the official SST test-set release and a new test set derived from the same sources and labeling methods as for `bakeoff_dev`.

For this bakeoff, you'll evaluate your original system from the above homework problem on these test sets. Our metric will be the mean of the macro-F1 values, which weights both datasets equally despite their differing sizes.

The central requirement for your system is that you have define a `predict_one` method for it that maps a text (str) directly to a label prediction – one of 'positive', 'negative', 'neutral'. If you used `sst.experiment` with `vectorize=True`, then the following function (for `softmax_experiment`) will be easy to adapt – you probably just need to change the variable `softmax_experiment` to the variable for your experiment output.

In [ ]:
def predict_one_softmax(text):
    # Singleton list of feature dicts:
    feats = [softmax_experiment['phi'](text)]
    # Vectorize to get a feature matrix:
    X = softmax_experiment['train_dataset']['vectorizer'].transform(feats)
    # Standard sklearn `predict` step:
    preds = softmax_experiment['model'].predict(X)
    # Be sure to return the only member of the predictions,
    # rather than the singleton list:
    return preds[0]

If you used an RNN like the one we demoed above, then featurization is a bit more straightforward:

In [ ]:
def predict_one_rnn(text):
    # List of tokenized examples:
    X = [rnn_experiment['phi'](text)]
    # Standard `predict` step on a list of lists of str:
    preds = rnn_experiment['model'].predict(X)
    # Be sure to return the only member of the predictions,
    # rather than the singleton list:
    return preds[0]

The following function is used to create the bakeoff submission file. Its arguments are your `predict_one` function and an output filename (str).

In [ ]:
def create_bakeoff_submission(
        predict_one_func,
        output_filename='cs224u-sentiment-bakeoff-entry.csv'):

    bakeoff_test = sst.bakeoff_test_reader(SST_HOME)
    sst_test = sst.test_reader(SST_HOME)
    bakeoff_test['dataset'] = 'bakeoff'
    sst_test['dataset'] = 'sst3'
    df = pd.concat((bakeoff_test, sst_test))

    df['prediction'] = df['sentence'].apply(predict_one_func)

    df.to_csv(output_filename, index=None)

Thus, for example, the following will create a bake-off entry based on `predict_one_softmax`:

In [ ]:
# This check ensure that the following code only runs on the local environment only.
# The following call will not be run on the autograder environment.
if 'IS_GRADESCOPE_ENV' not in os.environ:
    pass
    create_bakeoff_submission(predict_one_softmax)

This creates a file `cs224u-sentiment-bakeoff-entry.csv` in the current directory. That file should be uploaded as-is. Please do not change its name.

Only one upload per team is permitted, and you should do no tuning of your system based on what you see in our bakeoff prediction file – you should not study that file in anyway, beyond perhaps checking that it contains what you expected it to contain. The upload function will do some additional checking to ensure that your file is well-formed.

People who enter will receive the additional homework point, and people whose systems achieve the top score will receive an additional 0.5 points. We will test the top-performing systems ourselves, and only systems for which we can reproduce the reported results will win the extra 0.5 points.

Late entries will be accepted, but they cannot earn the extra 0.5 points.

## Submission Instruction

Review and follow the [Homework and bake-off code: Formatting guide](hw_formatting_guide.ipynb).
Please do not change the file name as described below.

Submit the following files to Gradescope:

- `hw_sentiment.ipynb` (this notebook)
- `cs224u-sentiment-bakeoff-entry.csv` (bake-off output)
